### ISSUE TITLE
SDF generation does not seem to work correctly for Noto fonts

### ISSUE THREAD

**Issue by Bulmanator** [nothings/stb#659]
Generating SDF values using the [Noto Font Family](https://www.google.com/get/noto/help/cjk/) (specifically the NotoSansCJK.ttc font file) produces incorrect results.
I have also tried [Source Han Sans](https://github.com/adobe-fonts/source-han-sans) and this also produces incorrect result. I think Noto uses the CJK glyphs from Source Han Sans so this problem must stem from that instead of Noto itself.
Incorrect values shown below (all images produced by writing out the data with stbi_write_png):

- The letter 'C' with Noto:
![noto_c](https://user-images.githubusercontent.com/8584482/45704214-397ba000-bb6e-11e8-9b78-110891061a5a.png)
- The Kanji character '小' (Codepoint 0x5C0F) with Noto:
![noto_kanji](https://user-images.githubusercontent.com/8584482/45704314-7778c400-bb6e-11e8-9956-d44abe62df82.png)

However, when using the built in Windows font Yu Gothic, the results are perfectly fine:
- The letter 'C' with Yu Gothic:
![yugoth_c](https://user-images.githubusercontent.com/8584482/45704507-f66dfc80-bb6e-11e8-9b5c-6f158aa27a00.png)
- The Kanji character '小' with Yu Gothic:
![yugoth_kanji](https://user-images.githubusercontent.com/8584482/45704511-fbcb4700-bb6e-11e8-8cc1-47265f98b424.png)

It can also produced packed atlases with stbtt_PackFontRanges perfectly fine with Noto.

The code used to generate the SDF glyph:
```
s32 sdf_width, sdf_height, sdf_xoff, sdf_yoff;
s32 glyph = stbtt_FindGlyphIndex(&font_info, 'C');
u8 *sdf_bitmap = (u8 *) stbtt_GetGlyphSDF(&font_info, stbtt_ScaleForPixelHeight(&font_info, 128),
       glyph, 5, 180, 36, &sdf_width, &sdf_height, &sdf_xoff, &sdf_yoff);
```

**Comment by wheybags** [comment]
Can confirm this is still happening on latest master.

**Comment by posila** [comment]
The issue is that stbtt_GetGlyphSDF doesn't handle STBTT_vcubic vertex type.
It turns out getting the closest point on cubic bezier curve is not trivial problem, but there are some solvers for this: https://github.com/nrtaylor/CubicSplineClosestPoint, https://github.com/erich666/GraphicsGems/blob/master/gems/NearestPoint.c
But it might be easier and/or faster to convert the curve into linear segments and use those to generate SDF.

**Comment by nothings** [comment]
Yeah, makes sense, I think the vcubic support was probably added after the SDF support. Note we solve quadratics directly (requires solving a cubic), so the non-trivial solution to the cubic curve (with a fifth-degree polynomial solver) isn't impossible, but yeah, it probably makes more sense to tesselate to quadratics or line segments.

**Comment by rygorous** [comment]
Distance to a cubic requires solving a fifth-order polynomial though. This would require some decidedly non-trivial changes.

**Comment by zewt** [comment]
Spent some time trying to figure out what I was doing wrong before finding this, would be nice to have a note in the stbtt_GetGlyphSDF docs for this limitation.

**Comment by bytepatch** [comment]
Has anyone worked out a solution for this?

**Comment by Jannes1000Orks** [comment]
Hey!
I did a very quick and !!!VERY!!! dirty fix this morning. I do not recommed using this the way it is. But for future reference and if anyone wants to pick up where I left off:
`		int num_verts = stbtt_GetGlyphShape(info, glyph, &verts);

		//CUSTOM FIXING CODE STARTS HERE
		int num_extraverts = 0;
		for (i = 0, j = num_verts - 1; i < num_verts; j = i++)
		{
			if (verts[i].type == STBTT_vcubic)
			{
				int num_points = 0;
				stbtt__point points[4096];
				float Flatness = 0.35f / scale;
				float objspace_flatness_squared = Flatness * Flatness;
				stbtt__tesselate_cubic(points, &num_points, verts[j].x, verts[j].y,
					verts[i].cx, verts[i].cy,
					verts[i].cx1, verts[i].cy1,
					verts[i].x, verts[i].y,
					objspace_flatness_squared, 0);
				num_extraverts += num_points;
			}
		}
		if (num_extraverts > 0)
		{
			stbtt_vertex* verts2 = (stbtt_vertex*)STBTT_malloc((num_verts+num_extraverts)*sizeof(stbtt_vertex), info->userdata);
			int vc = 0;
			for (i = 0, j = num_verts - 1; i < num_verts; j = i++)
			{
				if (verts[i].type == STBTT_vcubic)
				{
					int num_points = 0;
					stbtt__point points[4096];
					float Flatness = 0.35f / scale;
					float objspace_flatness_squared = Flatness * Flatness;
					stbtt__tesselate_cubic(points, &num_points, verts[j].x, verts[j].y,
						verts[i].cx, verts[i].cy,
						verts[i].cx1, verts[i].cy1,
						verts[i].x, verts[i].y,
						objspace_flatness_squared, 0);

					for (int k = 0; k < num_points; ++k)
					{
						verts2[vc].x = points[k].x;
						verts2[vc].y = points[k].y;
						verts2[vc].type = STBTT_vline;
						vc++;
					}
				}
				else
				{
					verts2[vc] = verts[i];
					vc++;
				}
			}
			STBTT_free(verts, info->userdata);
			verts = verts2;
			num_verts += num_extraverts;
		}

		//CUSTOM FIXING CODE ENDS HERE

		data = (unsigned char*)STBTT_malloc(w * h, info->userdata);`

I essentially used the tesselation code to convert the cubic curve to a bunch of line segments. The main issue is that it is hard to predict how many new points are going to be inserted. I just used a buffer of 4k points. But no idea what is actually needed in which cases. In theory the tesselation function can add 2^16 points per cubic vertex. Sooo....

But i renders the glyphs somewhat nice, while retaining the properties of the stbtt_GetGlyphSDF function.
